app/services/assignment_service.py

```diff
@@ -118,29 +131,76 @@ async def get_assignment_by_name(self, name: str) -> AssignmentModel:
+
+        if "directory_path" in update_fields:
+            assignment.directory_path = update_fields["directory_path"]
+
+        if "available_date" in update_fields:
+            assignment.available_date = update_fields["available_date"]
+        
+        if "due_date" in update_fields:
+            assignment.due_date = update_fields["due_date"]
+
+        if assignment.available_date >= assignment.due_date:
+            raise AssignmentDueBeforeOpenException
+
         self.session.commit()
+
+        dispatch(ModifyAssignmentCrudEvent(assignment=assignment, modified_fields=list(update_fields.keys())))
+
         return assignment
     
-    async def update_assignment_directory_path(self, assignment: AssignmentModel, directory_path: str) -> AssignmentModel:
-        assignment.directory_path = directory_path
-        assignment.last_modified_date = func.current_timestamp()
-        self.session.commit()
-        return assignment
+    # Get the earliest time at which the given assignment is available
+    async def get_earliest_available_date(self, assignment: AssignmentModel) -> datetime | None:
+        if assignment.available_date is None: return None
+        earliest_deferral = self.session.query(ExtraTimeModel.deferred_time) \
+            .filter_by(assignment_id=assignment.id) \
+            .order_by(ExtraTimeModel.deferred_time) \
+            .first()
+        
+        return assignment.available_date + (earliest_deferral if earliest_deferral is not None else timedelta(0))
     
-    async def update_assignment_available_date(self, assignment: AssignmentModel, available_date: datetime) -> AssignmentModel:
-        assignment.available_date = available_date
-        assignment.last_modified_date = func.current_timestamp()
-        self.session.commit()
-        return assignment
+    # Gets the latest time at which the given assignment closes
+    async def get_latest_due_date(self, assignment: AssignmentModel) -> datetime | None:
+        if assignment.due_date is None: return None
+        latest_time = self.session.query(ExtraTimeModel.extra_time) \
+            .filter_by(assignment_id=assignment.id) \
+            .order_by(ExtraTimeModel.extra_time.desc()) \
+            .first()
+        
+        return assignment.due_date + (latest_time.extra_time if latest_time.extra_time is not None else timedelta(0))
     
-    async def update_assignment_due_date(self, assignment: AssignmentModel, due_date: datetime) -> AssignmentModel:
-        assignment.due_date = due_date
-        assignment.last_modified_date = func.current_timestamp()
-        self.session.commit()
-        return assignment
+    async def get_master_notebook_name(self, assignment: AssignmentModel) -> str:
+        return self._compute_master_notebook_name(assignment.name)
+    
+    @staticmethod
+    def _compute_master_notebook_name(self, assignment_name: str) -> str:
+        return f"{ assignment_name }-prof.ipynb"
```

Comment:
I've actually changed it to just {name}.ipynb in my new branch, and then the student notebook is {profnotebookname}-student.ipynb.

Also, the master notebook is private and not added to the repo/visible to students. This is because the master notebook contains the test cases